localization/ar_tag_based_localizer/src/ar_tag_based_localizer_core.cpp

```diff
@@ -0,0 +1,276 @@
+  */
+  it_ = std::make_unique<image_transport::ImageTransport>(shared_from_this());
+
+  /*
+    Subscribers
+  */
+  image_sub_ = it_->subscribe("~/input/image", 1, &ArTagBasedLocalizer::image_callback, this);
+  cam_info_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(
+    "~/input/camera_info", 1,
+    std::bind(&ArTagBasedLocalizer::cam_info_callback, this, std::placeholders::_1));
+
+  /*
+    Publishers
+  */
+  using geometry_msgs::msg::PoseWithCovarianceStamped;
+  auto qos = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_default));
+  qos.reliable();
+  qos.transient_local();
+  image_pub_ = it_->advertise("~/debug/result", 1);
+  pose_pub_ =
+    this->create_publisher<PoseWithCovarianceStamped>("~/output/pose_with_covariance", qos);
+
+  RCLCPP_INFO(this->get_logger(), "Setup of ar_tag_based_localizer node is successful!");
+  return true;
+}
+
+void ArTagBasedLocalizer::image_callback(const sensor_msgs::msg::Image::ConstSharedPtr & msg)
+{
+  if ((image_pub_.getNumSubscribers() == 0) && (pose_pub_->get_subscription_count() == 0)) {
+    RCLCPP_DEBUG(this->get_logger(), "No subscribers, not looking for ArUco markers");
+    return;
+  }
+
+  if (!cam_info_received_) {
+    RCLCPP_DEBUG(this->get_logger(), "No cam_info has been received.");
+    return;
+  }
+
+  builtin_interfaces::msg::Time curr_stamp = msg->header.stamp;
+  cv_bridge::CvImagePtr cv_ptr;
+  try {
+    cv_ptr = cv_bridge::toCvCopy(*msg, sensor_msgs::image_encodings::RGB8);
+  } catch (cv_bridge::Exception & e) {
+    RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
+    return;
+  }
+
+  cv::Mat in_image = cv_ptr->image;
+
+  // detection results will go into "markers"
+  std::vector<aruco::Marker> markers;
+
+  // ok, let's detect
+  detector_.detect(in_image, markers, cam_param_, marker_size_, false);
+
+  // for each marker, draw info and its boundaries in the image
+  for (const aruco::Marker & marker : markers) {
+    tf2::Transform tf_cam_to_marker = right_to_left_ * aruco_marker_to_tf2(marker);
+
+    geometry_msgs::msg::TransformStamped tf_cam_to_marker_stamped;
+    tf2::toMsg(tf_cam_to_marker, tf_cam_to_marker_stamped.transform);
+    tf_cam_to_marker_stamped.header.stamp = curr_stamp;
+    tf_cam_to_marker_stamped.header.frame_id = camera_frame_;
+    tf_cam_to_marker_stamped.child_frame_id = "detected_marker_" + std::to_string(marker.id);
+    tf_broadcaster_->sendTransform(tf_cam_to_marker_stamped);
+
+    geometry_msgs::msg::PoseStamped pose_cam_to_marker;
+    tf2::toMsg(tf_cam_to_marker, pose_cam_to_marker.pose);
+    pose_cam_to_marker.header.stamp = curr_stamp;
+    pose_cam_to_marker.header.frame_id = std::to_string(marker.id);
+    publish_pose_as_base_link(pose_cam_to_marker);
+
+    // drawing the detected markers
+    marker.draw(in_image, cv::Scalar(0, 0, 255), 2);
+  }
+
+  // draw a 3d cube in each marker if there is 3d info
+  if (cam_param_.isValid() && marker_size_ != -1) {
```

Comment:
Can `marker_size_` be `-1` ? If so, please describe the situation.

Reply:
It can be set by the user in the config, but it should be detected in the constructor, so the condition have been removed. 
https://github.com/autowarefoundation/autoware.universe/pull/4347/commits/141b91d611422ea510888f37c414fe3c8d785716